File: bin/plot_ranking_reliability.py
#!/usr/bin/env python3
"""
plot_ranking_reliability.py

Scatter plot of win_fraction vs mean_f1_across_studies from rankings_best.tsv.
Cell types in the upper-right are both high-performing and consistently best
across studies.  Faceted by key level.

Usage:
    python plot_ranking_reliability.py \
        --input rankings_best.tsv \
        --outdir ranking_reliability
"""

import argparse
import os

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.lines import Line2D
import numpy as np
import pandas as pd
import seaborn as sns

import sys
sys.path.insert(0, os.path.dirname(__file__))
from plot_utils import KEY_ORDER, set_pub_style


DEFAULT_INPUT = "rankings_best.tsv"

# Key level colors
KEY_COLORS = {
    "subclass": "#e41a1c",
    "class": "#377eb8",
    "family": "#4daf4a",
    "global": "#984ea3",
}

# Size range for n_studies encoding
SIZE_MIN = 30
SIZE_MAX = 250


def parse_args():
    parser = argparse.ArgumentParser(
        description="Win fraction vs F1 scatter plot"
    )
    parser.add_argument("--input", default=DEFAULT_INPUT,
                        help="Path to rankings_best.tsv")
    parser.add_argument("--outdir", default="ranking_reliability",
                        help="Output directory")
    return parser.parse_args()


def main():
    args = parse_args()
    os.makedirs(args.outdir, exist_ok=True)
    set_pub_style()

    df = pd.read_csv(args.input, sep="\t")
    print(f"Loaded {len(df)} rows from {args.input}")

    keys = [k for k in KEY_ORDER if k in df["key"].unique()]
    n_keys = len(keys)

    if n_keys == 0:
        print("No data. Exiting.")
        return

    # --- Faceted scatter: one panel per key ---
    n_cols = min(n_keys, 2)
    n_rows = (n_keys + n_cols - 1) // n_cols
    fig, axes = plt.subplots(n_rows, n_cols, figsize=(7 * n_cols, 6 * n_rows),
                              squeeze=False)

    # Global n_studies range for sizing
    ns_min = df["n_studies"].min()
    ns_max = df["n_studies"].max()

    for idx, key in enumerate(keys):
        r, c = divmod(idx, n_cols)
        ax = axes[r, c]
        key_df = df[df["key"] == key].copy()
        color = KEY_COLORS.get(key, "#333333")

        # Compute sizes from n_studies
        if ns_max > ns_min:
            sizes = SIZE_MIN + (key_df["n_studies"] - ns_min) / (ns_max - ns_min) * (SIZE_MAX - SIZE_MIN)
        else:
            sizes = (SIZE_MIN + SIZE_MAX) / 2

        ax.scatter(
            key_df["mean_f1_across_studies"],
            key_df["win_fraction"],
            s=sizes, c=color, alpha=0.7,
            edgecolors="white", linewidth=0.5,
        )

        # Label points
        for _, row in key_df.iterrows():
            ax.annotate(
                row["label"], (row["mean_f1_across_studies"], row["win_fraction"]),
                fontsize=6, alpha=0.8,
                xytext=(4, 4), textcoords="offset points",
            )

        # Reference lines
        ax.axhline(0.5, color="gray", linestyle="--", linewidth=0.5, alpha=0.4)
        ax.axvline(0.5, color="gray", linestyle="--", linewidth=0.5, alpha=0.4)

        ax.set_xlim(-0.05, 1.05)
        ax.set_ylim(-0.05, 1.15)
        ax.set_xlabel("Mean F1 Across Studies")
        ax.set_ylabel("Win Fraction")
        ax.set_title(key.capitalize(), fontsize=14, fontweight="bold")

    # Hide unused axes
    for idx in range(n_keys, n_rows * n_cols):
        r, c = divmod(idx, n_cols)
        axes[r, c].set_visible(False)

    # --- Size legend ---
    legend_handles = []
    for ns_val, lbl in [(ns_min, f"n={ns_min}"), (ns_max, f"n={ns_max}")]:
        if ns_max > ns_min:
            sz = SIZE_MIN + (ns_val - ns_min) / (ns_max - ns_min) * (SIZE_MAX - SIZE_MIN)
        else:
            sz = (SIZE_MIN + SIZE_MAX) / 2
        legend_handles.append(
            Line2D([0], [0], marker="o", color="w", markerfacecolor="gray",
                   markeredgecolor="white", markersize=np.sqrt(sz),
                   label=lbl)
        )
    fig.legend(
        handles=legend_handles, loc="lower center",
        ncol=len(legend_handles), fontsize=9, frameon=False,
        bbox_to_anchor=(0.5, -0.03), title="N Studies",
    )

    fig.suptitle("Reliability vs Performance", fontsize=16, fontweight="bold")
    fig.tight_layout(rect=[0, 0.04, 1, 0.96])

    outpath = os.path.join(args.outdir, "ranking_reliability.png")
    fig.savefig(outpath, bbox_inches="tight", dpi=200)
    plt.close(fig)
    print(f"Saved {outpath}")
    print(f"\nDone. Figures in {args.outdir}/")


if __name__ == "__main__":
    main()

File: bin/plot_ranking_reliability.R
#!/usr/bin/env Rscript
# Reliability vs Performance Scatter Plot
#
# Scatter plot of win_fraction vs mean_f1_across_studies from rankings_best.tsv.
# Cell types in the upper-right are both high-performing and consistently best
# across studies. Faceted by taxonomy key level. Labels placed with ggrepel to
# avoid overlap.
#
# Usage:
#   Rscript plot_ranking_reliability.R \
#     --input rankings_best.tsv \
#     --outdir ranking_reliability

library(argparse)
library(dplyr)
library(readr)
library(ggplot2)
library(ggrepel)

# -- Constants ----------------------------------------------------------------

KEY_ORDER <- c("subclass", "class", "family", "global")

KEY_COLORS <- c(
  subclass = "#e41a1c",
  class    = "#377eb8",
  family   = "#4daf4a",
  global   = "#984ea3"
)

# -- Main ---------------------------------------------------------------------

main <- function() {
  parser <- ArgumentParser(
    description = "Win fraction vs F1 scatter plot (reliability vs performance)"
  )
  parser$add_argument("--input", default = "rankings_best.tsv",
                      help = "Path to rankings_best.tsv")
  parser$add_argument("--outdir", default = "ranking_reliability",
                      help = "Output directory")
  args <- parser$parse_args()

  dir.create(args$outdir, showWarnings = FALSE, recursive = TRUE)

  df <- read_tsv(args$input, show_col_types = FALSE)
  cat(sprintf("Loaded %d rows from %s\n", nrow(df), args$input))

  keys <- intersect(KEY_ORDER, unique(df$key))
  if (length(keys) == 0) {
    cat("No matching key levels found. Exiting.\n")
    quit(status = 0)
  }

  df <- df %>%
    filter(key %in% keys) %>%
    mutate(key = factor(key, levels = keys))

  # Normalise n_studies to [2, 8] point-size range
  ns_min <- min(df$n_studies, na.rm = TRUE)
  ns_max <- max(df$n_studies, na.rm = TRUE)
  if (ns_max > ns_min) {
    df <- df %>%
      mutate(pt_size = 2 + (n_studies - ns_min) / (ns_max - ns_min) * 6)
  } else {
    df <- df %>% mutate(pt_size = 5)
  }

  n_cols <- min(length(keys), 4)
  n_rows <- ceiling(length(keys) / n_cols)

  p <- ggplot(df, aes(x = mean_f1_across_studies, y = win_fraction)) +
    geom_hline(yintercept = 0.5, linetype = "dashed",
               color = "grey60", linewidth = 0.4) +
    geom_vline(xintercept = 0.5, linetype = "dashed",
               color = "grey60", linewidth = 0.4) +
    geom_point(aes(size = n_studies, color = key), alpha = 0.75) +
    geom_text_repel(
      aes(label = label, color = key),
      size          = 3.0,
      max.overlaps  = Inf,
      box.padding   = 0.4,
      point.padding = 0.3,
      segment.size  = 0.3,
      segment.color = "grey50",
      show.legend   = FALSE
    ) +
    facet_wrap(~ key, ncol = n_cols,
               labeller = labeller(key = tools::toTitleCase)) +
    scale_color_manual(values = KEY_COLORS, guide = "none") +
    scale_size_continuous(
      name   = "N Studies",
      range  = c(2, 8),
      breaks = scales::pretty_breaks(n = 3)
    ) +
    scale_x_continuous(limits = c(-0.05, 1.05),
                       breaks = scales::pretty_breaks(n = 5)) +
    scale_y_continuous(limits = c(-0.05, 1.15),
                       breaks = scales::pretty_breaks(n = 5)) +
    labs(
      x     = "Mean F1 Across Studies",
      y     = "Win Fraction",
      title = "Reliability vs Performance"
    ) +
    theme_bw(base_size = 13) +
    theme(
      plot.title       = element_text(face = "bold", hjust = 0),
      plot.background  = element_rect(fill = "white", color = NA),
      panel.background = element_rect(fill = "white", color = NA),
      legend.position  = "bottom"
    )

  outpath <- file.path(args$outdir, "ranking_reliability.png")
  ggsave(outpath, p,
         width  = 5 * n_cols,
         height = 5.5 * n_rows + 0.5,
         dpi    = 200,
         bg     = "white")
  cat(sprintf("Saved %s\n", outpath))
  cat(sprintf("\nDone. Figures in %s/\n", args$outdir))
}

main()
